### PR TITLE
Rteqs/subsampling

### PR DESCRIPTION
**Pull request by rteqs**

### CODE REVIEW

File: runtime/mount/subsample.py [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+            )
+        )
+
+        x = "x_dense_rank" if x_is_categorical else x
+        y = "y_dense_rank" if y_is_categorical else y
+
+        trace_data = conn.sql(
+            f"""
+            select
+                distinct on ({distinct_cols})
```
Comment (maximsmol) [inline runtime/mount/subsample.py:221]: should this be `group by` instead? unsure

File: runtime/mount/subsample.py [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+            """
+        ).set_alias(f"trace_{i}")
+
+        agg_expr = ""
+        for col, val in [
+            (f"min({x}) as min_x", min_x),
+            (f"max({x}) as max_x", max_x),
+            (f"min({y}) as min_y", min_y),
+            (f"max({y}) as max_y", max_y),
+            # extremas
```
Comment (maximsmol) [inline runtime/mount/subsample.py:234]:
```suggestion
            # extrema
```
it's latin (singular would be `extremum`)

File: runtime/mount/subsample.py [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+            excluded_columns.append("y_dense_rank")
+
+        exclude_str = ", ".join(excluded_columns)
+
+        cell_size = 3
+        max_occupancy = 2
+
+        # todo(rteqs): handle categorical axis
+        min_max = trace_data.aggregate(agg_expr).set_alias("min_max")
+        trace_data = (
```
Comment (maximsmol) [inline runtime/mount/subsample.py:267]: ok mostly not looking at this

File: runtime/mount/subsample.py [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+    if not is_latest:
+        # todo(rteqs): figure out how to convert plotly fig to duckdb table
+        # df
+        # conn.register(key, df)
+        conn.execute(
+            """
+            insert into
+                plots_faas_catalog
+            values
+                ($name, $generation, null)
```
Comment (maximsmol) [inline runtime/mount/subsample.py:379]: specify column names in insert and do not add `null`s and other default values

File: runtime/mount/subsample.py [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+
+    if not is_latest:
+        url = await get_presigned_url(f"latch://{ldata_node_id}.node")
+        conn.read_csv(url).to_table(f"ldata_{ldata_node_id}")
+        conn.execute(
+            """
+            insert into
+                plots_faas_catalog
+            values
+                ($name, null, $last_modified_time)
```
Comment (maximsmol) [inline runtime/mount/subsample.py:347]: specify column names in insert and do not add `null`s and other default values

File: runtime/mount/subsample.py [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+            on conflict
+                (name)
+            do update
+            set
+                generation = $generation
+            """,
+            parameters={"name": key, "generation": cur_gen},
+        )
+
+    # convert fig to downsample config
```
Comment (maximsmol) [inline runtime/mount/subsample.py:389]: this should be `todo`?

File: runtime/mount/subsample.py [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+    config: PlotConfig = {"traces": [{"x": "x", "y": "y"}]}
+    relations = downsample(conn, key, config)
+
+
+async def downsample_df(
+    conn: DuckDBPyConnection, key: str, df: DataFrame, config: PlotConfig, cur_gen: int
```
Comment (maximsmol) [inline runtime/mount/subsample.py:395]: use `in_memory_{key}` as the table name to avoid conflicts

**Review by maximsmol — CHANGES_REQUESTED** [review]

File: runtime/mount/kernel.py [hunk 8]
```diff
@@ -763,8 +759,16 @@ async def send(self, msg: object) -> None:
         # print("[kernel] >", msg)
         await self.conn.send(msg)
 
-    async def send_plot_data(self, plot_id: str, key: str) -> None:
+    async def send_plot_data(
+        self, plot_id: str, key: str, config: PlotConfig | None = None
+    ) -> None:
+        if config is not None and config == self.plot_configs.get(plot_id):
```
[the remark this answers is not in the document]
Comment (rteqs) [reply runtime/mount/kernel.py:765]: ok

**Pull request merged**